### DISCUSSION TITLE
How to get dependency_override to work to skip authentication

### DISCUSSION THREAD

**Discussion by Technisiums** [fastapi/fastapi#10118] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [x] I commit to help with one of those options 👆

### Description

I'm trying to bypass the dependency injection for authentication for testing an endpoint, but it doesn't work anyway, I have tried following
- Monkeypathing
- adding a dependency override for individual routers
- adding global dependency override (as in code right now)

I don't have any idea what I'm doing wrong, even I tried to take help from ChatGPT, but all of its solutions don't work, and I always get 401 response as unauthorized.

auth.py
```python
oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")

def get_current_user(token: Annotated[str, Depends(oauth2_scheme)]):
    credentials_exception = HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Could not validate credentials",
        headers={"WWW-Authenticate": "Bearer"},
    )
    try:
        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
        username: str = payload.get("sub")
        if username is None:
            raise credentials_exception
        token_data = TokenData(username=username)
    except JWTError:
        raise credentials_exception
    user = get_user(service_users, username=token_data.username)
    if user is None:
        raise credentials_exception
    return user


def get_current_active_user(
        current_user: Annotated[User, Depends(get_current_user)]
):
    if current_user.disabled:
        raise HTTPException(status_code=400, detail="Inactive user")
    return current_user

```

healthcheck_router.py

```python
router = APIRouter()


@router.get('/healthcheck', status_code=status.HTTP_200_OK)
def health_check(current_user: Annotated[UserBase, Depends(get_current_active_user)]):
    return {"status": "OK"}

```

main.py
```python
from fastapi import FastAPI
from app.routers import auth, healthcheck, persons
from fastapi_sqlalchemy import DBSessionMiddleware
from app.config.settings import settings
app = FastAPI()
app.add_middleware(DBSessionMiddleware,db_url=settings.SQLALCHEMY_DATABASE_URI)

app.include_router(auth.router)
app.include_router(healthcheck.router)
app.include_router(persons.router)

```

conftest.py
```python
@pytest.fixture(scope="module")
def unauthenticated_test_app():
    app.dependency_overrides = {
        "get_current_active_user": UserBase(full_name='Test User', username='test.user', email='test.user@app.com',
                                            disabled=False)
    }
    client = TestClient(app)
    yield client

```

test.py
```python
def test_healthcheck_router(unauthenticated_test_app):

    response = unauthenticated_test_app.get(f"/healthcheck")
    assert response.status_code == 200

```

### Operating System

macOS

### Operating System Details

13.5

### FastAPI Version

0.101.1

### Pydantic Version

2.2.0

### Python Version

3.9.6

**Comment by chrisK824** [comment, marked as answer]
You need to literally replace the callables, not pass the function's name. Try this for your `conftest.py` instead:

```
from auth import get_current_active_user

def get_current_active_user_test():
    return UserBase(full_name='Test User', username='test.user', email='test.user@app.com', disabled=False)

@pytest.fixture(scope="module")
def unauthenticated_test_app():
    app.dependency_overrides[get_current_active_user] = get_current_active_user_test
    client = TestClient(app)
    yield client
```
